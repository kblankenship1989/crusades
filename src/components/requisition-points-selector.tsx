import React from 'react';
import {View, Button} from 'react-native';
import {Icon, Card} from 'react-native-elements';
import {getColorScheme} from '../helpers/getColorScheme';
import {appStyles} from '../../styles';

export type IncrementorDecrementor = 1| -1;

export type RequisitionPointsSelectorProps = {
    currentPoints: number
    updateRequisitionPoints: (changeValue: IncrementorDecrementor) => void
};

const colorScheme = getColorScheme();
const styles = appStyles(colorScheme);


const getIconToRender = (index : number, currentPoints : number) : JSX.Element => {
    return  (
        <Icon
            size={18}
            type={'material-community'}
            name={index < currentPoints ? 'skull' : 'skull-outline'}
            key={`requisition-points-${index}`}
        />
    );
};

export const RequisitionPointsSelector = ({currentPoints, updateRequisitionPoints} : RequisitionPointsSelectorProps) : JSX.Element => {
    return (
        <Card
            title={'Requisition Points'}
        >
            <View
                style={styles.row}
                accessibilityLabel={`Current Points: ${currentPoints}/5`}
            >
                <Button
                    title={'-'}
                    disabled={currentPoints===0}
                    onPress={() => updateRequisitionPoints(-1)}
                />
                {getIconToRender(0, currentPoints)}
                {getIconToRender(1, currentPoints)}
                {getIconToRender(2, currentPoints)}
                {getIconToRender(3, currentPoints)}
                {getIconToRender(4, currentPoints)}
                <Button
                    title={'+'}
                    disabled={currentPoints === 5}
                    onPress={() => updateRequisitionPoints(1)}
                />
            </View>
        </Card>
    );
};
