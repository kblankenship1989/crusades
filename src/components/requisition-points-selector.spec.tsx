import React from 'react';
import '@testing-library/jest-native/extend-expect';
import {RequisitionPointsSelector, RequisitionPointsSelectorProps} from './requisition-points-selector';
import {render} from '@testing-library/react-native';

describe('Given the Requisition points selector', () => {
    it('should render correctly', () => {
        const testProps : RequisitionPointsSelectorProps = {
            currentPoints: 3,
            updateRequisitionPoints: jest.fn()
        };
        const component = (<RequisitionPointsSelector
            {...testProps}
        />);
        expect(render(component).toJSON()).toMatchSnapshot();
    });

    describe('and a requisition point value of 5', () => {
        it('should disable the incrementor button', async () => {
            const testProps : RequisitionPointsSelectorProps = {
                currentPoints: 5,
                updateRequisitionPoints: jest.fn()
            };
            const component = (<RequisitionPointsSelector
                {...testProps}
            />);

            const {
                getByText,
                queryByA11yLabel
            } = render(component);

            expect(await getByText('+')).toBeDisabled();
            expect(await getByText('-')).toBeEnabled();
            expect(await queryByA11yLabel('Current Points: 5/5')).toBeTruthy();
        });
    });

    describe('and a requisition point value of 0', () => {
        it('should disable the decrementor button', async () => {
            const testProps : RequisitionPointsSelectorProps = {
                currentPoints: 0,
                updateRequisitionPoints: jest.fn()
            };
            const component = (<RequisitionPointsSelector
                {...testProps}
            />);

            const {
                getByText,
                queryByA11yLabel
            } = render(component);

            expect(await getByText('+')).toBeEnabled();
            expect(await getByText('-')).toBeDisabled();
            expect(await queryByA11yLabel('Current Points: 0/5')).toBeTruthy();
        });
    });

    describe('and a requisition point value between 1 and 4', () => {
        it('should enable both the incrementor and decrementor buttons', async () => {
            const testProps : RequisitionPointsSelectorProps = {
                currentPoints: 4,
                updateRequisitionPoints: jest.fn()
            };
            const component = (<RequisitionPointsSelector
                {...testProps}
            />);

            const {
                getByText,
                queryByA11yLabel
            } = render(component);

            expect(await getByText('+')).toBeEnabled();
            expect(await getByText('-')).toBeEnabled();
            expect(await queryByA11yLabel('Current Points: 4/5')).toBeTruthy();
        });
    });
});
