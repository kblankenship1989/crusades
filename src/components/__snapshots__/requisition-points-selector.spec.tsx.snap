// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Requisition points selector should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderColor": "#e1e8ee",
      "borderWidth": 1,
      "margin": 15,
      "marginBottom": 0,
      "padding": 15,
      "shadowColor": "rgba(0,0,0, .2)",
      "shadowOffset": Object {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#43484d",
          "fontSize": 17.5,
          "fontWeight": "bold",
          "marginBottom": 15,
          "textAlign": "center",
        }
      }
      testID="cardTitle"
    >
      <Text>
        Requisition Points
      </Text>
    </Text>
    <View
      accessibilityLabel="Current Points: 3/5"
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            disabled={false}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            -
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            disabled={false}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            +
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
